lexos/models/kmeans_model.py

```diff
@@ -0,0 +1,475 @@
+        contains cluster numbers and the second column contains document names,
+        the rest columns contain the coordinates of the files.
+        """
+        # Get reduced data.
+        reduced_data = self._get_reduced_data()
+        # Get file names.
+        labels = [self._id_temp_label_map[file_id]
+                  for file_id in self._doc_term_matrix.index.values]
+
+        # Initialize the table with proper headers.
+        return pd.DataFrame(data={
+            "Cluster #": [index + 1 for index in k_means_index],
+            "Document": labels,
+            "X-Coordinate": reduced_data[:, 0],
+            "Y-Coordinate": reduced_data[:, 1]
+        })
+
+    def _get_3d_frame(self, k_means_index: List[int]) -> pd.DataFrame:
+        """Generate 3 dimensional table result for K-Means analysis.
+
+        :param k_means_index: the cluster result of K-Means analysis.
+        :return: A pandas data frame with four columns.The first column
+        contains cluster numbers and the second column contains document names,
+        the rest columns contain the coordinates of the files.
+        """
+        # Get reduced data.
+        reduced_data = self._get_reduced_data()
+        # Get file names.
+        labels = [self._id_temp_label_map[file_id]
+                  for file_id in self._doc_term_matrix.index.values]
+
+        # Initialize the table with proper headers.
+        return pd.DataFrame(data={
+            "Cluster #": [index + 1 for index in k_means_index],
+            "Document": labels,
+            "X-Coordinate": reduced_data[:, 0],
+            "Y-Coordinate": reduced_data[:, 1],
+            "Z-Coordinate": reduced_data[:, 2]
+        })
+
+    @staticmethod
+    def _get_voronoi_background(k_means: KMeans,
+                                reduced_data: np.ndarray) -> go.Heatmap:
+        """Plot polygons around each cluster.
+
+        The function first finds the decision boundary of the entire graph.
+        Then it calculates boundaries for each polygon around the clusters.
+        :param k_means: The fitted KMeans object.
+        :param reduced_data: PCA reduced two dimensional data.
+        :return: A plotly heat map object that contains all polygons.
+        """
+        # Find list of x, y coordinates.
+        x_value, y_value = reduced_data[:, 0], reduced_data[:, 1]
+
+        # Find min, max for x and then calculate bounds and step.
+        x_min, x_max = x_value.min(), x_value.max()
+        x_low_bound = x_min - (x_max - x_min) / 5
+        x_up_bound = x_max + (x_max - x_min) / 5
+        # Increase 200 will make lines smoother.
+        x_step = (x_up_bound - x_low_bound) / 200
+
+        # Find min, max for y and then calculate bounds and step.
+        y_min, y_max = y_value.min(), y_value.max()
+        y_low_bound = y_min - (y_max - y_min) / 5
+        y_up_bound = y_max + (y_max - y_min) / 5
+        y_step = (y_up_bound - y_low_bound) / 200
```

Comment:
I think you can create a helper function for these line, since you are repeating yourself here.

Reply:
If you don't have time, this is also fine.

Reply:
I wanted to move on to something else.. So I left a ```TODO``` for this.